### ISSUE TITLE
export the XDG user variables as an .environment file

### ISSUE THREAD

**Issue by xenoterracide** [systemd/systemd#20427]
**Is your feature request related to a problem? Please describe.**
A *LOT* of applications require you to [set these variables](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) to consider putting things in the right place. It's frustrating. As an application developer, it's also frustrating to have to write the code to check both.

**Describe the solution you'd like**

Just export them as a file in the first place https://www.freedesktop.org/software/systemd/man/environment.d.html#/dev/null and ship it in the right place for the distribution.

I'm not certain if you can use `systemctl edit` to override, so I suspect that's necessary, unless you set them in `--user` too and that will always override.

**Describe alternatives you've considered**

obviously, I can just do this myself, which isn't hard, but that only benefits me

**The systemd version you checked that didn't have the feature you are asking for**

current docs ;) also

```
❯ systemctl --version                                                                           # ~
systemd 248 (248.7-1-manjaro)
+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```

**Comment by poettering** [comment]
I am sorry, but I cannot parse this. "Export XDG variables as .environment" file? what is that supposed to mean?

**Comment by xenoterracide** [comment]
export the variables 

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

via this functionality

https://www.freedesktop.org/software/systemd/man/environment.d.html

*by default*

since your name is on both of those, I assume you know what they say

**Comment by xenoterracide** [comment]
> A LOT of applications require you to set these variables to consider putting things in the right place

Because application developers *aren't* following the spec, and often refuse to fix. In general, as said at the top, it's more work to actually follow the spec  than if the variables were set all the time. In the cases where they're partially supporting it by letting you use the variables, having them always available would simply make their apps behave correctly. Or more correctly, assuming they aren't stuffing everything in one directory, which is also common.

Basically, this would fix mediocre actors, for everyone (theoretically).

**Comment by poettering** [comment]
I am sorry, but I don't think this makes sense. No distro I know actually sets any of these vars by default, I have never done so myself beyond testing, and this never was an issue. systemd is really not the place to work around hypothetic bugs in other projects. Bugs that I don't think even exist.

I am sorry, but no.

**Comment by xenoterracide** [comment]
I decided just to do this in my own home directory, but I had problems, both of these should have worked according to the [documentation](https://www.freedesktop.org/software/systemd/man/environment.d.html#Configuration%20Format)  as I read it, `${HOME}`, `"${HOME}"` and `$HOME` should be interchange, but the first 2 didn't export. But only the last one did, is this a bug?

this didn't work after a relogin
```
❯ cat ~/.config/environment.d/xdg.conf
XDG_DATA_HOME=${HOME}/.local/share
XDG_CONFIG_HOME=${HOME}/.config
XDG_STATE_HOME=${HOME}/.local/state
XDG_CACHE_HOME=${HOME}/.cache
❯ env | grep XDG
XDG_SESSION_TYPE=wayland
XDG_RUNTIME_DIR=/run/user/1000
XDG_DATA_DIRS=/home/xeno/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share/:/usr/share/:/var/lib/snapd/desktop
XDG_SESSION_DESKTOP=gnome
XDG_CURRENT_DESKTOP=GNOME
XDG_SESSION_CLASS=user
XDG_MENU_PREFIX=gnome-
```

but this did

```
❯ cat ~/.config/environment.d/xdg.conf
XDG_DATA_HOME=$HOME/.local/share
XDG_CONFIG_HOME=$HOME/.config
XDG_STATE_HOME=$HOME/.local/state
XDG_CACHE_HOME=$HOME/.cache
❯ env | grep XDG
XDG_STATE_HOME=/home/xeno/.local/state
XDG_SESSION_TYPE=wayland
XDG_CACHE_HOME=/home/xeno/.cache
XDG_RUNTIME_DIR=/run/user/1000
XDG_DATA_DIRS=/home/xeno/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share/:/usr/share/:/var/lib/snapd/desktop
XDG_SESSION_DESKTOP=gnome
XDG_CURRENT_DESKTOP=GNOME
XDG_DATA_HOME=/home/xeno/.local/share
XDG_CONFIG_HOME=/home/xeno/.config
XDG_SESSION_CLASS=user
XDG_MENU_PREFIX=gnome-
```
---
I don't actually expect you to reconsider here, just something I need to respond

> myself beyond testing, and this never was an issue. systemd is really not the place to work around hypothetic bugs in other projects.

In my experience, support and conformance to the spec is sporadic, incomplete, and or just wrong. You could say that's the app developer's problem, but I would sometimes argue it's a bad spec, especially if there's a lot of bad actors. This could also be caused by more independent developers who don't actually realize there's a spec. Of course, the only problem this causes in practice seems to be annoyance at having a bunch of files in places where they shouldn't be, and still having a cluttered $HOME. It does not often actually cause things to not work (I've never experienced that). I wouldn't be surprised, that some things don't use the environment variables if set, like my app.

 I could dig through my GitHub for the issues… why not, and these are just the ones I found that I created on GitHub. That doesn't include other repository and bug systems. Not to mention, one of my apps doesn't implement it correctly yet.

* https://github.com/vercel/hyper/issues/5859
* https://github.com/ohmyzsh/ohmyzsh/issues/10058
* https://github.com/nvm-sh/nvm/issues/2409

I'm sure if you went through this list, you'd find more

https://github.com/issues?q=is%3Aopen+is%3Aissue+archived%3Afalse+xdg+spec

**Comment by ssipos90** [comment]
@xenoterracide mind you, XDG_* have default paths if they aren't set/exported, as per the spec.
